common/lib/dependabot/dependency_updaters/README.md

```diff
@@ -0,0 +1,63 @@
+# Dependency Updater
+
+A *dependency updater* is responsible for updating a dependent package
+according to a set of requirements.
+It provides a minimal interface for languages to add Dependabot support
+by delegating responsibilities to a package manager for that language.
+
+## Public API
+
+Each `Dependabot::DependencyUpdaters` class inherits from `Base` 
+and implements the following methods:
+
+### `#update(dependency:, requirements:)`
+
+Updates a dependency according to the provided requirements.
+This method is called within the root directory of a project.
+
+#### Arguments
+
+| Argument       | Type                     | Description                      |
+| -------------- | ------------------------ | -------------------------------- |
+| `dependency`   | `Dependabot::Dependency` | The package dependency to update |
+| `requirements` | `VersionRange`           | The updated version constraints  |
+
+#### Return
+
+Returns `true` if the dependency was successfully updated 
+to satisfy the new requirements.
```

Comment:
I think it could be super useful if this returned some data like:

- the dependencies that were updated with their `from` and `to` values, wether or not they were top-level or transitive maybe?
- files that were changed

Reply:
I'm with you. Why settle for `true` when we could return a more useful "truthy" value, right?

My only concern would be adding complexity for a hypothetical benefit. How would we use "from" and "to" values?

As far as file changes, this method is performed within the context of a Git worktree in a filesystem, so we can easily calculate that afterward.

Reply:
> My only concern would be adding complexity for a hypothetical benefit. How would we use "from" and "to" values?

In the PR title/description for example

> As far as file changes, this method is performed within the context of a Git worktree in a filesystem, so we can easily calculate that afterward.

Slightly worried we might end up picking up unintended changes, but we can figure out and learn if that's actually a problem or me just being paranoid

Reply:
> In the PR title/description for example

Let's game that out: How important is it for a PR to say "Update `vulnerable` to version 2.0.1"? Or would something like "Fix <CVE> on `vulnerable` 2.0.0" be sufficient with the raw diff? (I have no idea! That sounds like a product / UX question, so we can ask the relevant stakeholders)

> Slightly worried we might end up picking up unintended changes, but we can figure out and learn if that's actually a problem or me just being paranoid.

Conversely, taking the results verbatim from the package manager operating on the filesystem ensures that the PR doesn't leave out details. For example, a Go update that updates checksum values and a `vendor/` subdirectory, or a theoretical package manager that automatically generates an SBOM.

Reply:
> How important is it for a PR to say "Update vulnerable to version 2.0.1"?

From a product perspective not sure but we can ask around.

We do tend to use that information to show metadata about the PR like relevant changelog entries, commits between the versions etc etc. I think that information is very useful. 

I'm more concerned about not having to update all the code that's downstream from these changes that expects that information to be present. 

My thinking is that the more we're able to isolate the changes to a portion of the system the easier it'll be to integrate, and also to move existing ecosystems over.

Reply:
What do you think about taking a progressive enhancement approach? 

Plugins with a basic level of support return a "truthy" return value to indicate success and a "falsey" return value (`nil` or `false` (or raising an exception)) to indicate failure. The resulting PR may not have all of the information to populate a full description, but it works all the same.

Plugins that want to support additional functionality  integration could construct a response object with additional information to fill out the description with changelog entries, commits, and all the rest.

You're more familiar with that side of the codebase. Do you see any opportunities to take a tiered / relaxed / progressive approach like this to generating PRs?

@exvuma Thoughts about the product impact here? How important do you think it is for all PRs generated by Dependabot to have the same level of information for all ecosystems (trading off against the ease of supporting more ecosystems)?

Reply:
> Do you see any opportunities to take a tiered / relaxed / progressive approach like this to generating PRs?

I think it'll be pretty hard to integrate with the existing PR creation code if we don't have that information. Do you foresee many problems getting the before/after versions of a dependency? It seems like it should be somewhat doable to grab those, I'd assume the package manager would definitely have that information in order to be able to resolve the dependency tree.

For example, in the case of `npm` you could do:

```bash
$ npm ls etag --json
{
  "version": "1.0.0",
  "name": "name",
  "dependencies": {
    "etag": {
      "version": "1.8.1",
      "resolved": "https://registry.npmjs.org/etag/-/etag-1.8.1.tgz"
    }
  }
}
```

Reply:
> I think it'll be pretty hard to integrate with the existing PR creation code if we don't have that information.

👍 

I like the push to progressively enhancing the PR metadata, also good if sources are down but I also have alarm bells around the state of the codebase for PR creation. Branch creation and rebasing would likely break if we omitted the to version for example as we use this to create unique branch names per pm/package/version. For example: https://github.com/dependabot/dependabot-core/tree/dependabot/pip/python/helpers/hashin-0.17.0 checked here https://github.com/dependabot/dependabot-core/blob/dependabot/pip/python/helpers/hashin-0.17.0/common/lib/dependabot/pull_request_creator/github.rb#L43